VerilogCodingStyle.md

```diff
@@ -2291,6 +2292,109 @@ always_comb begin
 end
 ```
 
+### Functions and Tasks
+
+The following section applies to synthesizable RTL only.
+
+***In synthesizable RTL use of functions is allowed, provided they are declared
+automatic. Tasks should not be used.***
+
+Functions must be declared outside of a class in either a package or inside a
+module. A package is appropriate where the function relates to other definitions
+in the package and could be useful to multiple modules (even if its currently
+only used by one). A module is appropriate where the function specifically
+relates to the internals of that module.
+
+Functions should aim to conceptually represent a reusable block of combinational
+logic.
+
+Storage types must be explicitly declared for all arguments and the function
+return value.
+
+Do not use `input`, `output`, `inout`, or `ref` on function arguments.
```

Comment:
Is there a reason to avoid these (except maybe `ref`)? The LRM doesn't indicate that these render the function unsynthesizable (AFAIK).

Reply:
I feel like `ref` should be excluded, and `inout` may be a candidate we may want to exclude because of style reasons (avoid tristate constructs on-chip). But for more than just one return value, `output` could be useful...

Reply:
Although I agree that functions should be kept simple and just return one value if possible...

Reply:
This is a style guide after all, so I think it is a good idea to limit it to good practice,
which IMO is all inputs and just the one output. So if we're going with this particular
guidance, I would restate it as follows, since `input` per se isn't a problem.

  Do not use `output`, `inout`, or `ref` on function arguments. All functions should
  only consume inputs and produce the one function output. `input` is the default and
  is not required on the function arguments.

Then probably add this to your {.bad} examples

&#x1f44e;
```systemverilog {.bad}
function automatic logic [2:0] foo(input [2:0] a, input [2:0] b, output [2:0] c);
  c = a - b;
  return a + b;
endfunction
```

Reply:
recommend this modification.